core/blockchain.go

```diff
@@ -514,17 +514,21 @@ func (self *BlockChain) Recovery(from int, increment int) (checkpoint uint64) {
 // assumes that the chain manager mutex is held.
 func (self *BlockChain) LoadLastState(dryrun bool) error {
 
+	self.mu.Lock()
 	// recoverOrReset checks for recoverable block data.
 	// If no recoverable block data is found, plain Reset() is called.
 	// If safe blockchain data exists (so head block was wrong/invalid), blockchain db head
 	// is set to last available safe checkpoint.
 	recoverOrReset := func() error {
```

Comment:
It's error prone to call unlock in so many places, and it's easy to forget about it when extending code. Different option is: use `defer self.mu.Unlock()` after `self.mu.Lock()` (line 517), and then in `recoverOrReset`, call 
```
self.mu.Unlock()
defer self.mu.Lock()
``` 
It's rather not standard approach, but may be practical in this case.
It's just a crazy idea, not "I insist on this change" comment.

Reply:
I agree with you completely; I'm just not sure adding

```
func (self *BlockChain) LoadLastState(dryrun bool) error {

	self.mu.Lock()
	defer self.mu.Unlock()

	// recoverOrReset checks for recoverable block data.
	// If no recoverable block data is found, plain Reset() is called.
	// If safe blockchain data exists (so head block was wrong/invalid), blockchain db head
	// is set to last available safe checkpoint.
	recoverOrReset := func() error {
		glog.V(logger.Warn).Infoln("Checking database for recoverable block data...")

		recoveredHeight := self.Recovery(1, 100)

		self.mu.Unlock()
		defer self.mu.Lock()

		if recoveredHeight == 0 {
			glog.V(logger.Warn).Infoln("WARNING: No recoverable data found, resetting to genesis.")
			return self.Reset()
```
will allow the returned functions to access the `mu`...?

Reply:
Please be welcome to push the changes you have in mind if I have it wrong

Reply:
As per [spec](https://golang.org/ref/spec#Defer_statements) deferred function is executed after function returns so this has to work, Proof-of-Concept: https://play.golang.org/p/5rHVeNas6k

Reply:
Roger that, thanks :) See 83c4614